src/librustc_lint/unused.rs

```diff
@@ -270,47 +270,52 @@ impl UnusedParens {
                                 msg: &str,
                                 struct_lit_needs_parens: bool) {
         if let ast::ExprKind::Paren(ref inner) = value.node {
-            let necessary = struct_lit_needs_parens &&
-                            parser::contains_exterior_struct_lit(&inner);
-            if !necessary {
-                let span_msg = format!("unnecessary parentheses around {}", msg);
-                let mut err = cx.struct_span_lint(UNUSED_PARENS,
-                                                  value.span,
-                                                  &span_msg);
-                // Remove exactly one pair of parentheses (rather than naïvely
-                // stripping all paren characters)
-                let mut ate_left_paren = false;
-                let mut ate_right_paren = false;
-                let parens_removed = pprust::expr_to_string(value)
-                    .trim_matches(|c| {
-                        match c {
-                            '(' => {
-                                if ate_left_paren {
-                                    false
-                                } else {
-                                    ate_left_paren = true;
-                                    true
-                                }
-                            },
-                            ')' => {
-                                if ate_right_paren {
-                                    false
-                                } else {
-                                    ate_right_paren = true;
-                                    true
-                                }
-                            },
-                            _ => false,
-                        }
-                    }).to_owned();
-                err.span_suggestion_short_with_applicability(
-                    value.span,
-                    "remove these parentheses",
-                    parens_removed,
-                    Applicability::MachineApplicable
-                );
-                err.emit();
+            if struct_lit_needs_parens &&
+                    (parser::contains_exterior_struct_lit(&inner) ||
+                        if let ast::ExprKind::Break(..) = inner.node {
```

Comment:
Nit: I'd prefer to see this `if let` captured into a temporary value, or perhaps with a helper function. This is a pretty complex conditional! Maybe even break it into a `if`

Reply:
That'd be a pity; I just found a way to use a `return` instead of a boolean that would shorten the code considerably. But I can (and should) add a comment why we don't lint on `break` here.

Reply:
I don't follow what would be a pity...? I was just suggesting reformatting things

Reply:
Sorry, I've been busy lately. I'll revisit and rebase this later this week.